README.md
# LRSDAY

<p align="center">
  <img src="https://github.com/yjx1217/LRSDAY/blob/master/LRSDAY.logo.png" alt="LRSDAY logo" width="627" height="242"/>
</p>

**LRSDAY: Long-read Sequencing Data Analysis for Yeasts**

A highly transparent, automated and powerful computational framework for high-quality genome assembly and annotation.

[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)

## Description
Long-read sequencing technologies have become increasingly popular in genome projects due to their strengths in resolving complex genomic regions. As a leading model organism with small genome size and great biotechnological importance, the budding yeast, *Saccharomyces cerevisiae*, has many isolates currently being sequenced with long reads. However, analyzing long-read sequencing data to produce high-quality genome assembly and annotation remains challenging. Here we present LRSDAY, the first one-stop solution to streamline this process. LRSDAY can produce chromosome-level end-to-end genome assembly and comprehensive annotations for various genomic features (including centromeres, protein-coding genes, tRNAs, transposable elements and telomere-associated elements) that are ready for downstream analysis. Although tailored for *S. cerevisiae*, we designed LRSDAY to be highly modular and customizable, making it adaptable for virtually any eukaryotic organisms. 

![LRSDAY_flowchart](https://github.com/yjx1217/LRSDAY/blob/master/LRSDAY_flowchart.png)


## Citations
Jia-Xing Yue & Gianni Liti. (2018) Long-read sequencing data analysis for yeasts. *Nature Protocols*, 13:1213–1231. 

Jia-Xing Yue, Jing Li, Louise Aigrain, Johan Hallin, Karl Persson, Karen Oliver, Anders Bergström, Paul Coupland, Jonas Warringer, Marco Cosentino Lagomarsino, Gilles Fischer, Richard Durbin, Gianni Liti. (2017) Contrasting evolutionary genome dynamics between domesticated and wild yeasts. *Nature Genetics*, 49:913-924.

## Release history
* v1.7.2 Released on 2023/03/20
* v1.7.1 Released on 2023/01/28
* v1.7.0 Released on 2022/12/31
* v1.6.0 Released on 2019/10/03
* v1.5.0 Released on 2019/05/13
* v1.4.0 Released on 2019/03/21
* v1.3.1 Released on 2019/01/22
* v1.3.0 Released on 2018/11/13
* v1.2.0 Released on 2018/10/15
* v1.1.0 Released on 2018/07/11
* v1.0.0 Released on 2018/02/04

## License
LRSDAY itself is distributed under the MIT license. A number of LRSDAY's dependencies (e.g. CAP3, MAKER, GATK, blat, RepBase, etc) are under more restricted licenses, for which commerical use of the software needs to be discussed with the corresponding developers.


## Requirements
### Hardware, operating system and network
This protocol is designed for a desktop or computing server running an x86-64-bit Linux operating system. Multithreaded processors are preferred to speed up the process since many steps can be configured to use multiple threads in parallel. For assembling and analyzing the budding yeast genomes (genome size = ~12.5 Mb), at least 16 Gb of RAM and 100 Gb of free disk space are recomended. More disk space is needed if you want to use LRSDAY to do Nanopore basecalling. When adapted for other eukaryotic organisms with larger genome sizes, the RAM and disk space consumption will scale up, majorly during *de novo* genome assembly (performed by [Canu](https://github.com/marbl/canu) by default. Plese refer to [Canu’s manual](http://canu.readthedocs.io/en/latest/) for suggested RAM and disk space consumption for assembling large genomes. Stable Internet connection is required for the installation and configuration of LRSDAY as well as for retrieving the test data.


### Software or library requirements
* bash (https://www.gnu.org/software/bash/)
* bzip2 and libbz2-dev (http://www.bzip.org/)
* cmake v3.2 or newer (https://cmake.org/)
* gcc and g++ v4.9.1 or newer (https://gcc.gnu.org/)
* ghostscript (https://www.ghostscript.com)
* git (https://git-scm.com/)
* gnu make (https://www.gnu.org/software/make/)
* gzip (https://www.gnu.org/software/gzip/)
* java runtime environment (JRE) v1.8.0 (https://www.java.com)
* perl v5.12 or newer (https://www.perl.org/)
* python v2.7.9 or newer (https://www.python.org/)
* python-devel
* python v3.7 or newer (https://www.python.org/)
* python3-devel
* tar (https://www.gnu.org/software/tar/)
* unzip (http://infozip.sourceforge.net/UnZip.html)
* virtualenv v15.1.0 or newer (https://virtualenv.pypa.io)
* wget (https://www.gnu.org/software/wget/)
* zlib and zlib-devel (https://zlib.net/)



CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [1.7.2] - 2023-03-20
### Changed
- Better support for historical PacBio RSII reads.
- Cleaner TE output in the final annotation GFF3 file.

## [1.7.1] - 2023-01-28
### Fixed
- A bug leading to the loss of detailed TE subtype info in the final annotation GFF3 file.

## [1.7.0] - 2022-12-31
### Changed
- Better robunsteness for the dependency installation script.
- Functional enhancement for the basecalling, demultiplexing, and quality summary of ONT raw fast5 reads.
- Version updates for multiple genome assemblers.
- Functional enhancement for the mitochondrial genome assembly.
- Functional enhancement for the mitochondrial gene annotation.
- Functional enhancement for the nuclear gene annotation.
- Functional enhancement for the TE annotation.
- Functional enhancement for SGD_based orthology identification and tagging.
- Removal of support for the long-read-assembler: Ra.
- Removal of support for the long-read-polisher: MarginPolish.

## [1.6.0] - 2019-10-03
### Added
- Supports for canu-triobinning assembly mode for phased hybrid genome assembly.
- Supports for new alternative long-read-assemblers: Ra and Shasta.
- Supports for new alternative long-read-based polishers: Racon, Medaka, and MarginPolish.
- Supports for new alternative reference-genome-based scaffolder: RaGOO.

### Changed
- Better robustness for the dependency installation script.
- Better treatment for identifying mitochondrial contigs when they contain assembly gaps.
- Minor adjustment for a few task-specific directory names.
- Software version updates for a number of dependencies.
### Fixed
- A bug that will disrupt long-read-based assembly run when assembler="canu-flye", "canu-wtdbg2" or "canu-smartdenovo" because of deleting a non-existing file.

## [1.5.0] - 2019-05-13
### Added
- Supports for native ONT nanopore basecalling, demultiplexing, and read profile plotting.
- An lite-weight bash script for generating dotplot for any pairs of fasta files.
### Changed
- Better robustness for the dependency installation script.
- Reducing the requirements of setting output prefix by assuming the same prefix was used across different modules.
- Setting adjustment for long-read filtering and downsampling.
- Applying read filtering/downsampling for the testing example.
- Software version updates for a number of dependencies.
### Fixed
- A bug that might lead to missing gene annotation in certain genomic region.
- Typos in the installation script.
- Typos in the manual.

## [1.4.0] - 2019-03-21
### Changed
- Supports for multi-round assembly polishing using both long and short reads.
- Setting adjustment for long-read filtering and downsampling.
- Software version updates for a number of dependencies. 
### Fixed
- Compatibility issues due to recent version updates of conda an bioconda.
- Typos in the installation script.

## [1.3.1] - 2019-01-22
### Added
- A script for generated demultiplexed fastq reads based on nanopore's guppy demultiplexing summary file.
- A script for extracting all read IDs from the input fastq file.
- A script for calculating the sequence length of each enclosed sequences from the input fasta file.
### Changed
- Software version updates for a number of dependencies. Importantly, nanopolish was bumped up to v0.11.0 to support the new multi-fast5 nanopore reads.
- Downloading URL updates for testing data due to changes on the EBI data server. 
- Adopting picard tools' new commandline syntax.
### Fixed
- A bug that prevents correct parameter parsing when multiple customized canu parameters are specified.
- Minor bugs related with deleting intermediate files in special cases.

## [1.3.0] - 2018-11-13
### Added
- Support for one more alternative assembler: wtdbg2.
### Changed
- Substantially more automated installation/setup process.
- Software version or downloading URL updates for a number of dependencies.
### Fixed
- Bugs introduced due to changes made for file/parameter names in the LRSDAY.01.Long-read-based_Genome_Assembly.sh script when using some alternative assemblers.
- Mismatched step numbers and file names in the manual due to previous version changes.
- Typos in the manual.

## [1.2.0] - 2018-10-15
### Added
- Support for adapter trimming for Nanopore reads (via Porechop).
- Support for long-read filtering based on both quality and length (via Filtlong).
- Support for long-read-based polishing for PacBio and Nanopore reads (via Quiver/Arrow for PacBio reads and nanopolish for Nanopore reads).
- Support for the bax2bam format conversion for the PacBio RSII reads to make it compatible with PacBio's current SMRT pipeline.
- Support for dedicated mitochondrial gene annotation (via Mfannot).
### Changed
- Treat nuclear genome and mitochondrial genome separately in the annotation phase.
- Better robustness for various processing scripts.
- Software version or downloading URL updates for a number of dependencies.
### Fixed
- Typos in the manual.

## [1.1.0] - 2018-07-11
### Added
- This change log file: Changelog.md
- Support for customized parameter settings for Canu.
- Support for alternative assemblers: Flye and smartdenovo.
- The script subsampling_seqeunces.pl for downsampling input reads.
- Support for the report of exact numbers of SNP and INDEL corrections made based on the Illumina data.
### Changed
- Better resource management for the Illumina-based polishing step.
- Better robustness for the reference-based scaffolding step.
- Better performance and robustness for the mitochondrial genome assembly improvement step.
- Better robustness for the TE annotation step.
- Better robustness for the reference-based gene orthology identification step.
- Software version or downloading URL updates for a number of dependencies.

scripts/cds2protein.pl
#!/usr/bin/perl
use warnings;
use strict;
use Getopt::Long;

##############################################################
#  script: cds2protein.pl
#  author: Jia-Xing Yue (GitHub ID: yjx1217)
#  last edited: 2018.10.03
#  description: translate cds sequences into protein sequences
#  example: perl cds2protein.pl -i input.cds.fa(.gz) -t 1 -p prefix
#  See this link (https://www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprintgc.cgi) for the genetic code table options. Only "1" and "3" are supported so far. 
##############################################################

my ($input, $genetic_code_table, $prefix);
$genetic_code_table = 1;
GetOptions('input|i:s' => \$input, # input file
	   'genetic_code_table|t:s' => \$genetic_code_table, # "1" for the standard code and "3" for the yeast mitochondrial code.
	   'prefix|p:s' => \$prefix); # file name prefix for output files

my $input_fh = read_file($input);
my @input = ();
my %input = ();
parse_fasta_file($input_fh, \%input, \@input);
my $trimmed_cds_output = "$prefix.trimmed_cds.fa";
my $trimmed_cds_output_fh = write_file($trimmed_cds_output);
my $pep_output = "$prefix.pep.fa";
my $pep_output_fh = write_file($pep_output);
my $trimmed_cds_log = "$prefix.trimmed_cds.log";
my $trimmed_cds_log_fh = write_file($trimmed_cds_log);

my $pseudogene_output = "$prefix.manual_check.list";
my $pseudogene_output_fh = write_file($pseudogene_output);

foreach my $id (@input) {
    # print "id = $id\n";
    my @pseudogene_test = ();
    my $cds = uc $input{$id};
    $cds =~ tr/U/T/;
    if (($cds !~ /^ATG/) and ($cds =~ /(TAA|TAG|TGA)$/)) {
	push @pseudogene_test, "unexpected start codon based on standard genentic code;your selected code table is $genetic_code_table";
    } elsif (($cds =~ /^ATG/) and ($cds !~ /(TAA|TAG|TGA)$/)) {
	push @pseudogene_test, "unexpected stop codon based on standard genentic code;your selected code table is $genetic_code_table";
    } elsif (($cds !~ /^ATG/) and ($cds !~ /(TAA|TAG|TGA)$/)) {
	push @pseudogene_test, "unexpected start & end codons based on standard genentic code;your selected code table is $genetic_code_table";
    }
    my $cds_length = length $cds;
    my $length_test = $cds_length % 3;
    my $cds_trim = $cds;

    if ($length_test != 0) {
	push @pseudogene_test, "incorrect CDS length";
	my $cds_phase0 = substr $cds, 0, $cds_length - $length_test;
	my $pep_phase0 = translate($genetic_code_table, $cds_phase0);
	my $pep_phase0_stop_count = () = $pep_phase0 =~ /\*/g;
	my $cds_phase1 = substr $cds, 1, $cds_length - $length_test + 1;
	my $pep_phase1 = translate($genetic_code_table, $cds_phase1);
	my $pep_phase1_stop_count = () = $pep_phase1 =~ /\*/g;
	my $cds_phase2 = substr $cds, 2, $cds_length - $length_test + 2;
	my $pep_phase2 = translate($genetic_code_table, $cds_phase2);
	my $pep_phase2_stop_count = () = $pep_phase2 =~ /\*/g;
	if (($pep_phase0_stop_count <= $pep_phase1_stop_count) and ($pep_phase0_stop_count <= $pep_phase2_stop_count)) {
	    # print "my guess is phase 0, resulting a total of $pep_phase0_stop_count stop codons\n";
	    $cds_trim = $cds_phase0;
	    print $trimmed_cds_log_fh "incorrect_length: $id, phase_guess: 0\n";
	} elsif (($pep_phase1_stop_count <= $pep_phase0_stop_count) and ($pep_phase1_stop_count <= $pep_phase2_stop_count)) {
            # print "my guess is phase 1, resulting a total of $pep_phase1_stop_count stop codons\n";
            $cds_trim = $cds_phase1;
	    print $trimmed_cds_log_fh "incorrect_length: $id, phase_guess: 1\n";
	} elsif (($pep_phase2_stop_count <= $pep_phase0_stop_count) and ($pep_phase2_stop_count <= $pep_phase1_stop_count)) {
            # print "my guess is phase 2, resulting a total of $pep_phase2_stop_count stop codons\n";
            $cds_trim = $cds_phase2;
	    print $trimmed_cds_log_fh "incorrect_length: $id, phase_guess: 2\n";
	}
    }
    print $trimmed_cds_output_fh ">$id\n$cds_trim\n";
    my $protein = translate($genetic_code_table, $cds_trim);
    # get rid of the ending stop codon for translated protein sequences
    $protein =~ s/\*$//;
    print $pep_output_fh ">$id\n$protein\n";
    if (($protein =~ /\*/) and ($length_test == 0)) {
	push @pseudogene_test, "internal stop codon(s)";
    }
    if ((scalar @pseudogene_test) > 0) {
	my $pseudogene_test = join ";", @pseudogene_test;
	print $pseudogene_output_fh "$id\t$pseudogene_test\n";
    }
}


sub read_file {
    my $file = shift @_;
    my $fh;
    if ($file =~ /\.gz$/) {
        open($fh, "gunzip -c $file |") or die "can't open pipe to $file";
    } else {
        open($fh, $file) or die "can't open $file";
    }
    return $fh;
}

sub write_file {
    my $file = shift @_;
    my $fh;
    if ($file =~ /.gz$/) {
        open($fh, "| gzip -c >$file") or die "can't open $file\n";
    } else {
        open($fh, ">$file") or die "can't open $file\n";
    }
    return $fh;
}  


sub parse_fasta_file {
    my ($fh, $input_hashref, $input_arrayref) = @_;
    my $seq_name = "";
    while (<$fh>) {
	chomp;
	if (/^\s*$/) {
	    next;
	} elsif (/^\s*#/) {
	    next;
	} elsif (/^>(.*)/) {
	    $seq_name = $1;
	    push @$input_arrayref, $seq_name;
	    $$input_hashref{$seq_name} = "";
	} else {
	    $$input_hashref{$seq_name} .= $_;
	}
    }
}


sub translate {
    my ($genetic_code_table, $dna) = @_;
    my $protein = "";
    for (my $i = 0; $i < (length($dna) - 2); $i += 3) {
	$protein .= codon2aa($genetic_code_table, substr($dna, $i, 3));
    }
    return $protein;
}


sub codon2aa {
    my ($genetic_code_table, $codon) = @_;
    $codon = uc $codon;
    my %genetic_code = ();
    %{$genetic_code{"1"}} = (
	'TCA' => 'S',    # Serine
	'TCC' => 'S',    # Serine
	'TCG' => 'S',    # Serine
	'TCT' => 'S',    # Serine
	'TCN' => 'S',    # Serine
	'TTC' => 'F',    # Phenylalanine
	'TTT' => 'F',    # Phenylalanine
	'TTA' => 'L',    # Leucine
	'TTG' => 'L',    # Leucine
	'TAC' => 'Y',    # Tyrosine
	'TAT' => 'Y',    # Tyrosine
	'TAA' => '*',    # Stop
	'TAG' => '*',    # Stop
	'TGC' => 'C',    # Cysteine
	'TGT' => 'C',    # Cysteine
	'TGA' => '*',    # Stop
	'TGG' => 'W',    # Tryptophan
	'CTA' => 'L',    # Leucine
	'CTC' => 'L',    # Leucine
	'CTG' => 'L',    # Leucine
	'CTT' => 'L',    # Leucine
	'CTN' => 'L',    # Leucine
	'CCA' => 'P',    # Proline
	'CCC' => 'P',    # Proline
	'CCG' => 'P',    # Proline
	'CCT' => 'P',    # Proline
	'CCN' => 'P',    # Proline
	'CAC' => 'H',    # Histidine
	'CAT' => 'H',    # Histidine
	'CAA' => 'Q',    # Glutamine
	'CAG' => 'Q',    # Glutamine
	'CGA' => 'R',    # Arginine
	'CGC' => 'R',    # Arginine
	'CGG' => 'R',    # Arginine
	'CGT' => 'R',    # Arginine
	'CGN' => 'R',    # Arginine
	'ATA' => 'I',    # Isoleucine
	'ATC' => 'I',    # Isoleucine
	'ATT' => 'I',    # Isoleucine
	'ATG' => 'M',    # Methionine
	'ACA' => 'T',    # Threonine
	'ACC' => 'T',    # Threonine
	'ACG' => 'T',    # Threonine
	'ACT' => 'T',    # Threonine
	'ACN' => 'T',    # Threonine
	'AAC' => 'N',    # Asparagine
	'AAT' => 'N',    # Asparagine
	'AAA' => 'K',    # Lysine
	'AAG' => 'K',    # Lysine
	'AGC' => 'S',    # Serine
	'AGT' => 'S',    # Serine
	'AGA' => 'R',    # Arginine
	'AGG' => 'R',    # Arginine
	'GTA' => 'V',    # Valine
	'GTC' => 'V',    # Valine
	'GTG' => 'V',    # Valine
	'GTT' => 'V',    # Valine
	'GTN' => 'V',    # Valine
	'GCA' => 'A',    # Alanine
	'GCC' => 'A',    # Alanine
	'GCG' => 'A',    # Alanine
	'GCT' => 'A',    # Alanine
	'GCN' => 'A',    # Alanine
	'GAC' => 'D',    # Aspartic Acid
	'GAT' => 'D',    # Aspartic Acid
	'GAA' => 'E',    # Glutamic Acid
	'GAG' => 'E',    # Glutamic Acid
	'GGA' => 'G',    # Glycine
	'GGC' => 'G',    # Glycine
	'GGG' => 'G',    # Glycine
	'GGT' => 'G',    # Glycine
	'GGN' => 'G',    # Glycine
	'MGR' => 'R',    # Arginine
	'AAY' => 'N',    # Asparagine
	'GAY' => 'D',    # Aspartic Acid
	'TGY' => 'C',    # Cysteine
	'CAR' => 'Q',    # Glutamine
	'GAR' => 'E',    # Glutamic Acid
	'CAY' => 'H',    # Histidine
	'ATH' => 'I',    # Isoleucine
	'YTR' => 'L',    # Leucine
	'AAR' => 'K',    # Lysine
	'TTY' => 'F',    # Phenylalanine
	'TCN' => 'S',    # Serine
	'AGY' => 'S',    # Serine
	'ACN' => 'T',    # Threonine
	'TAY' => 'Y',    # Tyrosine
	'TAR' => '*',    # Stop
	'TRA' => '*',    # Stop
	'---' => '-'     # alignment gap
	);
    for (my $i = 2; $i <= 31; $i++) {
	%{$genetic_code{$i}} =  %{$genetic_code{"1"}};
	if ($i eq "2") {
	    $genetic_code{$i}{'AGA'} = '*'; # Stop
	    $genetic_code{$i}{'AGG'} = '*'; # Stop
	    $genetic_code{$i}{'AGR'} = '*'; # Stop
	    $genetic_code{$i}{'ATA'} = 'M'; # Methionine
	    $genetic_code{$i}{'TGA'} = 'W'; # Tryptophan
	} elsif ($i eq "3") {
	    $genetic_code{$i}{'ATA'} = 'M'; # Methionine
	    $genetic_code{$i}{'CTT'} = 'T'; # Threonine
	    $genetic_code{$i}{'CTC'} = 'T'; # Threonine
	    $genetic_code{$i}{'CTA'} = 'T'; # Threonine
	    $genetic_code{$i}{'CTG'} = 'T'; # Threonine
	    $genetic_code{$i}{'CTR'} = 'T'; # Threonine
	    $genetic_code{$i}{'CTY'} = 'T'; # Threonine
	    $genetic_code{$i}{'CTS'} = 'T'; # Threonine
	    $genetic_code{$i}{'CTW'} = 'T'; # Threonine
	    $genetic_code{$i}{'CTK'} = 'T'; # Threonine
	    $genetic_code{$i}{'CTM'} = 'T'; # Threonine
	    $genetic_code{$i}{'CTB'} = 'T'; # Threonine
	    $genetic_code{$i}{'CTD'} = 'T'; # Threonine
	    $genetic_code{$i}{'CTH'} = 'T'; # Threonine
	    $genetic_code{$i}{'CTN'} = 'T'; # Threonine
	    $genetic_code{$i}{'TGA'} = 'W'; # Trptophan	    
	} elsif ($i eq "4") {
	    $genetic_code{$i}{'TGA'} = 'W'; # Tryptophan
	} elsif ($i eq "5") {
	    $genetic_code{$i}{'AGA'} = 'S'; # Serine
	    $genetic_code{$i}{'AGG'} = 'S'; # Serine
	    $genetic_code{$i}{'AGR'} = 'S'; # Serine
	    $genetic_code{$i}{'ATA'} = 'M'; # Methionine
	    $genetic_code{$i}{'TGA'} = 'W'; # Tryptophan
	} elsif ($i eq "6") {
	    $genetic_code{$i}{'TAA'} = 'Q'; # Glutamine
	    $genetic_code{$i}{'TAG'} = 'Q'; # Glutamine
	    $genetic_code{$i}{'TAR'} = 'Q'; # Glutamine
	} elsif ($i eq "9") {
	    $genetic_code{$i}{'AAA'} = 'N'; # Asparagine
	    $genetic_code{$i}{'AGA'} = 'S'; # Serine
	    $genetic_code{$i}{'AGG'} = 'S'; # Serine
	    $genetic_code{$i}{'AGR'} = 'S'; # Serine
	    $genetic_code{$i}{'TGA'} = 'W'; # Tryptophan
	} elsif ($i eq "10") {
	    $genetic_code{$i}{'TGA'} = 'C'; # Cysteine
	} elsif ($i eq "12") {
	    $genetic_code{$i}{'CTG'} = 'S'; # Serine
	} elsif ($i eq "13") {
	    $genetic_code{$i}{'AGA'} = 'G'; # Glycine
	    $genetic_code{$i}{'AGG'} = 'G'; # Glycine
	    $genetic_code{$i}{'AGR'} = 'G'; # Glycine
	    $genetic_code{$i}{'ATA'} = 'M'; # Methionine
	    $genetic_code{$i}{'TGA'} = 'W'; # Tryptophan
	} elsif ($i eq "14") {
	    $genetic_code{$i}{'AAA'} = 'N'; # Asparagine
	    $genetic_code{$i}{'AGA'} = 'S'; # Serine
	    $genetic_code{$i}{'AGG'} = 'S'; # Serine
	    $genetic_code{$i}{'AGR'} = 'S'; # Serine
	    $genetic_code{$i}{'TAA'} = 'Y'; # Tyrosine
	    $genetic_code{$i}{'TGA'} = 'W'; # Tryptophan
	} elsif ($i eq "16") {
	    $genetic_code{$i}{'TAG'} = 'L'; # Leucine
	} elsif ($i eq "21") {
	    $genetic_code{$i}{'TGA'} = 'W'; # Tryptophan
	    $genetic_code{$i}{'ATA'} = 'M'; # Methionine
	    $genetic_code{$i}{'AGA'} = 'S'; # Serine
	    $genetic_code{$i}{'AGG'} = 'S'; # Serine
	    $genetic_code{$i}{'AGR'} = 'S'; # Serine
	    $genetic_code{$i}{'AAA'} = 'N'; # Asparagine
	} elsif ($i eq "22") {
	    $genetic_code{$i}{'TAG'} = 'L'; # Leucine
	} elsif ($i eq "24") {
	    $genetic_code{$i}{'AGA'} = 'S'; # Serine
	    $genetic_code{$i}{'AGG'} = 'K'; # Lysine
	    $genetic_code{$i}{'TGA'} = 'W'; # Tryptophan
	} elsif ($i eq "25") {
	    $genetic_code{$i}{'TGA'} = 'G'; # Glycine
	} elsif ($i eq "26") {
	    $genetic_code{$i}{'ATG'} = 'A'; # Alanine
	} elsif ($i eq "27") {
	    $genetic_code{$i}{'TAG'} = 'Q'; # Glutamine
	    $genetic_code{$i}{'TAA'} = 'Q'; # Glutamine
	    $genetic_code{$i}{'TAR'} = 'Q'; # Glutamine
	} elsif ($i eq "29") {
	    $genetic_code{$i}{'TAG'} = 'Y'; # Tyrosine
            $genetic_code{$i}{'TAA'} = 'Y'; # Tyrosine
            $genetic_code{$i}{'TAR'} = 'Y'; # Tyrosine
	} elsif ($i eq "30") {
	    $genetic_code{$i}{'TAG'} = 'E'; # Glutamic Acid
	    $genetic_code{$i}{'TAA'} = 'E'; # Glutamic Acid
	    $genetic_code{$i}{'TAR'} = 'E'; # Glutamic Acid
	} elsif ($i eq "31") {
	    $genetic_code{$i}{'TGA'} = 'W'; # Tryptophan
	}
    }
    
    if (exists $genetic_code{$genetic_code_table}{$codon}) {
        return $genetic_code{$genetic_code_table}{$codon};
    } else {
	#print STDERR "Bad codon \"$codon\"!!\n";
	return 'X';
    }
}

